### PR TITLE
[AutoDiff upstream] Add `differentiable_function` canonicalization.

### PR DESCRIPTION
**Pull request by dan-zheng**
Canonicalizes `differentiable_function` instructions by filling in missing
derivative function operands.

Derivative function emission rules, based on the original function value:

- `function_ref`: look up differentiability witness with the exact or a minimal
  superset derivative configuration. Emit a `differentiability_witness_function`
  for the derivative function.
- `witness_method`: emit a `witness_method` with the minimal superset derivative
  configuration for the derivative function.
- `class_method`: emit a `class_method` with the minimal superset derivative
  configuration for the derivative function.

If an *actual* emitted derivative function has a superset derivative
configuration versus the *desired* derivative configuration, create a "subset
parameters thunk" to thunk the actual derivative to the desired type.

For `differentiable_function` instructions formed from curry thunk applications:
clone the curry thunk (with type `(Self) -> (T, ...) -> U`) and create a new
version with type `(Self) -> @differentiable (T, ...) -> U`.

Progress towards TF-1211.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp:1143]: Note: subset parameters thunk tests are blocked by derivative function emitters, which will be upstreamed next.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp:171]: Todo: upstream [`differentiation_transform_diagnostics.swift`](https://github.com/apple/swift/blob/tensorflow/test/AutoDiff/downstream/differentiation_transform_diagnostics.swift), which tests `diagnoseUnsatisfiedRequirements`.

**Comment by dan-zheng** [comment]
@gottesmm: would you like to help review this patch, as a SILOptimizer code owner?

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Review by rxwei — APPROVED** [review]

**Comment by dan-zheng** [comment]
@swift-ci Please test Linux

**Comment by dan-zheng** [comment]
Merging to unblock progress. Happy to address any feedback later!

File: include/swift/SILOptimizer/Utils/Differentiation/Common.h
Comment (gottesmm) [inline include/swift/SILOptimizer/Utils/Differentiation/Common.h:39]: I have a number of problems with this. I am pretty sure we already have something like this.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp:1008]: If there is a common pattern here, maybe put the construction of this function onto a helper on SILOptFunctionBuilder? The hope with that was to hide the low level SILFunction creation functions in favor of higher level routines so we don't have to have huge invocations of SILFunctionType::get.

File: lib/SILOptimizer/Utils/Differentiation/Common.cpp
Comment (gottesmm) [inline lib/SILOptimizer/Utils/Differentiation/Common.cpp:42]: Is this really needed? Could we put this onto SILBuilder or something? Seems wrong to have such helper utilities.

File: lib/SILOptimizer/Utils/Differentiation/Common.cpp
Comment (gottesmm) [inline lib/SILOptimizer/Utils/Differentiation/Common.cpp:49]: This is void SILBuilder::emitDestructureValueOperation(SILLocation loc, SILValue operand,
                                     SmallVectorImpl<SILValue> &result);

File: lib/SILOptimizer/Utils/Differentiation/Common.cpp
Comment (gottesmm) [inline lib/SILOptimizer/Utils/Differentiation/Common.cpp:65]: This again doesn't feel like a method that has anything to do with differentiation.

**Review by gottesmm — COMMENTED** [review]
@dan-zheng some comments. Please fix:

Also, are these used anywhere but the differentiation pass? If not, we should just put it in the differentiation pass until we find multiple uses. A bunch of the helpers that you added need to be refactored. Please fix them. I may look at this again in a bit.

File: lib/SILOptimizer/Utils/Differentiation/Common.cpp
Comment (gottesmm) [reply lib/SILOptimizer/Utils/Differentiation/Common.cpp:42]: I would give it a name that makes it explicit that you are using a tuple. Otherwise, how do I know I am getting back a tuple?

**Comment by marcrasi** [comment]
> Also, are these used anywhere but the differentiation pass? If not, we should just put it in the differentiation pass until we find multiple uses.

Would moving all the contents of `include/SILOptimizer/Utils/Differentiation` and `lib/SILOptimizer/Utils/Differentiation` to a new directory named `lib/SILOptimizer/Mandatory/Differentiation` be an appropriate way to do this?

I would very much like to avoid moving everything into a single `.cpp` file because differentiation has many distinct bits of analysis and synthesis that are easier to comprehend separately. (Also having multiple files is really nice for incremental compilation time.)

**Comment by dan-zheng** [comment]
Thanks @gottesmm! I'll create a PR addressing your feedback, we can continue discussion there.

File: include/swift/SILOptimizer/Utils/Differentiation/Common.h
Comment (dan-zheng) [reply include/swift/SILOptimizer/Utils/Differentiation/Common.h:75]: This change seems important for correctness. It can probably help clean up this code:
https://github.com/apple/swift/blob/13d5a8addbe3605984edc4ce7c6cbf1f0649a9e1/lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp#L486-L497

I started looking into this but ran into some difficulties. The user above wants to insert at the end of a basic block (`SILBasicBlock::iterator`), but no appropriate constructor exists. Only `SILBuilderWithScope(SILInstruction *I, SILBuilderContext &C)` is recommended.

`SILBuilderWithScope(SILBasicBlock::iterator I, SILBuilderContext &C)` would work but doesn't exist. Can we add it, or are there other considerations?

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp:1008]: `SILOptFunctionBuilder` looks very minimal, so I'm not sure

File: lib/SILOptimizer/Utils/Differentiation/Common.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Utils/Differentiation/Common.cpp:42]: I renamed `joinElements` to `createTupleIfNotSingle` in https://github.com/apple/swift/pull/30970.  This helper is sufficiently ad-hoc that I wouldn't want to move it outside of differentiation to `SILBuilder`.

Feel free to give feedback on the PR!

File: lib/SILOptimizer/Utils/Differentiation/Common.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Utils/Differentiation/Common.cpp:49]: Addressing in https://github.com/apple/swift/pull/30970#discussion_r407013548, let's continue discussion there.

File: lib/SILOptimizer/Utils/Differentiation/Common.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Utils/Differentiation/Common.cpp:65]: Currently, this code currently emits `AdditiveArithmetic.zero`, but we plan to change it to emit `Differentiable.zeroTangentVector` sometime (TF-1005), so I'd prefer to keep it differentiation-specific.

Callers of this function could be deduped though: `PullbackEmitter::emitZeroIndirect` and `JVPEmitter::emitZeroIndirect`. I can look into that, but I'm not sure how to best use `SILBuilderContext &` since a `SILBasicBlock::iterator` is needed as an insertion point: https://github.com/apple/swift/pull/30818#discussion_r406975815

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp:1008]: @dan-zheng I wrote that code and the reason why it is minimal now is that I was trying to create the scaffolding for what I just mentioned. Just no one has had the time to fix it up old uses. That being said, we should not make the problem worse.

**Pull request merged**
